README.md
# snowdqq

changes.txt
1> have to deploy our application tonative app
2>add Project button 
3>remove the popups in the project page and should see a alternative solution for that
4>Snowdq complete demo according to the samir video reference  like suits adding and removing.
5> stored procedure for S3 bucket should be run same like adf.--will be provided by Samir

snowdq-basic_structure/main.py
import streamlit as st
import pandas as pd
import pandas as pd
from pages.Expectations import *
from streamlit_option_menu import option_menu
from streamlit_extras.switch_page_button import switch_page
from pandas import json_normalize
from utils import *
from statistics import mean
from streamlit_modal import Modal
from streamlit_extras.stylable_container import stylable_container

st.set_page_config(page_title="SnowDQ", page_icon="static/favicon.ico", layout="wide", initial_sidebar_state="collapsed")
navWithLogo()
# col1, col2 = st.columns((1,8))
# with col1:
#     pass
#     st.image('static/SnowDQ-LOGO.png')
# with col2:
#     page_selected = option_menu(
#             menu_title = None,
#             options = ['Projects', 'Suites', 'Expectations','Datasets'],
#             default_index = 0,
#             icons=['stack','bounding-box','file-earmark-text','folder2'],
#             menu_icon=None,
#             orientation='horizontal',
#             styles={
#             "container": {"padding": "0!important", "background-color": "#fafafa"},
#             # "icon": {"color": "orange", "font-size": "25px"},
#             "icon":{"display":"inline-block"},
#             "nav": {"background-color":"#f2f5f9"},
#             "nav-link": {"font-size": "15px",
#             "font-weight":"bold",
#             "color":"#1d4077",
#             "border-right":"0px solid #1d4077",
#             "border-left":"0px solid #1d4077",
#             "border-top":"0px solid #1d4077",
#             "border-bottom":"0px solid #1d4077",
#             "padding":" 10px",
#             "text-transform": "uppercase",
#             "border-radius":"5px",
#             "margin":"5px",
#             "--hover-color": "#e1e1e1"},
#             "nav-link-selected": {"background-color":"#1d4077", "color":"#ffffff"},
#             }
# )

# st.set_page_config(page_title="SnowDQ | Home", page_icon="static/favicon.ico", layout="wide")
projectsDf = load_data(st.secrets.DQ_TABLE.PROJECT)
projectsDf['MODIFIED_DATE'] = pd.to_datetime(projectsDf['MODIFIED_DATE'])
projectsDf = projectsDf.sort_values(by='MODIFIED_DATE', ascending=False)
dashboardDf = load_data(st.secrets.DQ_TABLE.VALIDATION_RESULTS)
distinctProject = projectsDf.drop_duplicates(subset='NAME')
 
col1,col2 = st.columns((2,8))
with col1:
    st.divider()
    cola,colb = st.columns((8,2))
    with cola:
        search = st.text_input("Search Project",label_visibility="visible", placeholder="Search Project...")
    with colb:
        buttons()
        createProject = st.button("\+", use_container_width=True)
        if "createProject" not in st.session_state:
                st.session_state["createProject"] = False
        if createProject:
                st.session_state["createProject"] = not st.session_state["createProject"]
        if st.session_state["createProject"]:
            createProjectcontainer()
            

    if search:
        df = distinctProject[distinctProject['NAME'].str.contains(search, case=False)]
        selected_row = st.radio("Select a Project:", df['NAME'])

    else:
        cola1, colb2 = st.columns((8,2))
        with cola1:
            selected_row = st.radio("Select a Project:", distinctProject['NAME'])
            print("selected", selected_row)
        with colb2:
            optionButton = st.button("⋮", key=selected_row, use_container_width=True)
            modal = Modal(key="Demo Key", title=" ")
            displayDf = projectsDf[projectsDf["NAME"] == selected_row]   #TODO check for search functionality
            if "optionButton" not in st.session_state:
                st.session_state["optionButton"] = False
            if optionButton:
                st.session_state["optionButton"] = not st.session_state["optionButton"]
            
            if st.session_state["optionButton"]:
                projectId = displayDf['ID'].to_string(index=False)
                print("id>>>>>>>>>>", projectId)
                optionContainer(optionButton, selected_row, projectId)


with col2:
    st.divider()
    if selected_row:
        vr = dashboardDf[dashboardDf['PROJECT_NAME'] == selected_row]
        dq_score= sum(vr['RESULT'].apply(lambda x: json.loads(x).get('success_percentage',0)))
        if dq_score == 0:
            dq_score = 0
        else:
            dq_score = round(dq_score/len(vr['RESULT'].apply(lambda x: json.loads(x).get('success_percentage',0))),2)
        total = sum(vr['RESULT'].apply(calculate_total))
        failed = sum(vr['RESULT'].apply(calculate_unexpected))
        passed = total -failed
        st.write(f"<h1>{displayDf['NAME'].to_string(index=False)}</h1>", unsafe_allow_html=True)
        st.write(f"<h4>{displayDf['DESCRIPTION'].to_string(index=False)}</h4>", unsafe_allow_html=True)
        project_dashboard(dqScore=dq_score, total=total, failed=failed, success=passed)
        st.divider()
        st.dataframe(vr, use_container_width=True,hide_index=True, height=360)
        st.divider()
    else:
        project_dashboard(dqScore=0, total=0, failed=0, success=0)
        st.divider()

snowdq-basic_structure/pages/Expectations.py
import streamlit as st
import pandas as pd
import json
from streamlit_option_menu import option_menu
from snowflake_client import load_data
from utils import *
from streamlit_extras.switch_page_button import switch_page

# st.set_page_config(page_title="SnowDQ | Expectations", page_icon="static/favicon.ico", layout="wide", initial_sidebar_state="collapsed")
navWithLogo()
expectationsDf = load_data(st.secrets.DQ_TABLE.EXPECTATIONS)

col1,col2 = st.columns((100,10))
with col1:
    search = st.text_input("Search Expectation",label_visibility="visible", placeholder="Search Expectation...")
with col2:
    items_per_page = st.selectbox("Rows Per Page",[10,25,50],label_visibility="visible")
    if items_per_page ==25 or items_per_page ==50:
        st.session_state.page = 1
if 'page' not in st.session_state:
    st.session_state.page = 1

if search:
    df = expectationsDf[expectationsDf['RULE'].str.contains(search, case=False)]
    page_data, total_pages = paginate_dataframe(df, st.session_state.page, items_per_page)
else:
    page_data, total_pages = paginate_dataframe(expectationsDf, st.session_state.page, items_per_page)

for index, row in page_data.iterrows():
    col1, col2, col3, col4, col5, col6, col7 = st.columns((4, 1, 1, 3, 4, 1,0.5))

    with col1:
        rule = page_data['RULE'][index]
        st.write(f'<span style="font-size:20px">{rule}</span>', unsafe_allow_html=True)
        st.write(page_data['DESCRIPTION'][index])
    
    with col2:
        st.write("Owner")
        # owner_circle(page_data["OWNER"][index][0])
        suite_owner_circle("GX")
    
    with col3:
        st.write("Category")
        category = page_data["CATEGORY"][index]
        st.write(f'<span style="font-size:24px">{category}</span>', unsafe_allow_html=True)
    
    with col4:
        st.write("Tags")
        tags = json.loads(page_data["TAGS"][index])
        if tags:
            c1,c2,c3 = st.columns((4,5,1))
            with c1:
                suite_rule_background(tags[0])
            with c2:
                suite_rule_background(tags[1])
        else:
            pass
            # suite_rule_background("None")

    with col5:
        st.write("Backend support")
        backend_support = json.loads(page_data["SUPPORTED_BACKEND"][index])
        if backend_support:
            c1,c2,c3 = st.columns((2,2,7))
            with c1:
                suite_rule_background(backend_support[0])
            with c2:
                if len(backend_support)>1:
                    suite_rule_background(backend_support[1])
            with c3:
                if len(backend_support)>2:
                    suite_rule_background(f"+{len(backend_support) -2}")
        else:
            pass
            # suite_rule_background("None")
        # suite_rule_background(backend_support)


    with col6:
        st.write("Status")
        status = page_data["STATUS"][index]
        success(status)

    with col7:
        click = st.button("⋮", key=f"button_{index}")  # Unique key for each button
        if click:
            html(click,row)
    
    st.markdown("""---""") 
col1, col2, col3= st.columns([40,150,20])
buttons()
if col1.button('Previous Page', key='prev_page'):
    if st.session_state.page > 1:
        st.session_state.page -= 1
        st.experimental_rerun()
if col3.button('Next Page', key='next_page'):
    if st.session_state.page < total_pages:
        st.session_state.page += 1
        st.experimental_rerun()

col2.write(f'Page {st.session_state.page}/{total_pages}')



utils.py
import pandas as pd
import streamlit as st
import json
from streamlit_modal import Modal
from streamlit_extras.stylable_container import stylable_container
from snowflake_client import *
from streamlit_extras.switch_page_button import switch_page


def paginate_dataframe(df, page_number, page_size):
    """
    Paginate a Pandas DataFrame and return the specified page.

    Args:
        df (pd.DataFrame): The DataFrame to paginate.
        page_number (int): The page number to retrieve (1-based index).
        page_size (int): The number of rows per page.

    Returns:
        pd.DataFrame: The specified page of the DataFrame.
        int: Total number of pages.
    """
    total_rows = len(df)
    total_pages = (total_rows + page_size - 1) // page_size  # Calculate total pages

    start_idx = (page_number - 1) * page_size
    end_idx = start_idx + page_size

    return df.iloc[start_idx:end_idx], total_pages

"""
Example usage:
page_number = 2  # Display the second page
page_size = 10  # 10 rows per page
page_data, total_pages = paginate_dataframe(df, page_number, page_size)
"""




def success(txt):
    # txt="Production"    
    htmlstr1=f"""<p style='background-color:#c7e9cc;
    color:#1fa531;
    font-size:14px;
    border-radius:30px;
    line-height:40px;
    padding-left:22px;
    opacity:0.6'>
    {txt}</style>
    <br></p>"""
    st.markdown(htmlstr1,unsafe_allow_html=True)
    
def suite_owner_circle(text):
    circle_html = f"""
    <!DOCTYPE html>
    <html>
    <head>
        <style>
            /* Define the styles for the circle */
            .circle {{
                width: 40px;
                height: 40px;
                background-color: #e6ecf3;
                border-radius: 50%;
                text-align: center;
                line-height: 40px;
                color: #08101c;
                font-size: 24px;
            }}
        </style>
    </head>
    <body>
        <!-- Create the circle and add text inside it -->
        <div class="circle">{text}</div>
    </body>
    </html>
    """
    st.markdown(circle_html,unsafe_allow_html=True)

def suite_rule_background(text):
    suite_rule_background = f"""
    <!DOCTYPE html>
<html>
<head>
    <style>
        /* Define the styles for the background */
        .background {{
            display: inline-block;
            background-color: #e6ecf3; /* Background color */
            padding: 10px; /* Padding to create space around the text */
            border-radius: 30px; /* Rounded corners */
            color: #08101c; /* Text color */
            font-size: 18px; /* Font size */
        }}
    </style>
</head>
<body>
    <!-- Create the text with a background -->
    <div class="background">{text}</div>
</body>
</html>
    """
    st.markdown(suite_rule_background,unsafe_allow_html=True)





def html(click):
    if click:
        right_popup_visible = True
    else:
        right_popup_visible = False
    
    right_popup_content = """
        <div style="position: fixed; z-index:9999; top: 0; right: 0; width:  1000px; height: 100%;background-color: white; box-shadow: -5px 0 5px rgba(0, 0, 0, 0.2);">
        <br>
        <br>
        <h5>Expectation Details</h5>
        <hr>
        <h5>expect_column_values_to_not_be_in_set</h5>
        <hr>
        <label>Description</label>
        <p>except column values to not be in set</p>
        <hr>
        <label> Tags</label>
        <p>[ "["core expectation", "column map expectation"]" ]<p>
        <hr>
        <p>Backend support()</P>
        <hr>
        <button style="position: absolute; bottom: 20px; right: 20px;" onclick="closePopup()">Cancel</button>
        </div>

    <script>
    function closePopup() {
        document.querySelector('#rightPopup').style.display = 'None';
    }
    </script>

    """

    
    if right_popup_visible:
        st.markdown(right_popup_content, unsafe_allow_html=True)


def buttons():
    button_style = """
    <style>
        .stButton>button {
            margin: 11px; /* Add a 5-pixel margin around the element */
            background-color: #e6ecf3;
            color: #08101c;
            border-radius: 5px;
            padding: 7px 20px;
            font-size: 16px;
            transition: background-color 0.3s ease, color 0.3s ease, transform 0.3s ease;
        }
        .stButton>button:hover {
            background-color: #021859;
            color: #e6ecf3;
            transform: translate(0px, -3px);
        }
    </style>
    """
    st.markdown(button_style, unsafe_allow_html=True)


def project_dashboard(dqScore=0, total=0, failed=0, success=0):
    html_code = f"""
    <style>
    /* Define the CSS for the boxes and container */
    .container {{
        display: flex;
        justify-content: space-between;
    }}
    .box {{
        background-color: #e6ecf3;
        width: 200px; /* Adjust the width as needed */
        height: 120px;
        padding: 20px;
        margin: 15px;
        border-radius: 15px;
    }}
    </style>
    <div class="container">
        <div class="box">
            <p>Average DQ Score</p>
            <p style="font-size: 24px;">{dqScore}</p>
        </div>
        <div class="box">
            <p>Total Records</p>
            <p style="font-size: 24px;">{total}</p>
        </div>
        <div class="box">
            <p>Failed</p>
            <p style="font-size: 24px; color: red;">{failed}</p>
        </div>
        <div class="box">
            <p>Success</p>
            <p style="font-size: 24px; color: green;">{success}</p>
        </div>
    </div>
    """

    st.markdown(html_code, unsafe_allow_html=True)




def calculate_total(result_json):
    # Initialize a variable to store the total
    total = 0
    # Convert the 'result_json' column value to a list of dictionaries
    result_list = json.loads(result_json).get('result_json', [])
    # Iterate through the list and calculate the total
    for item in result_list:
        result = item.get('result', {})
        total += result.get('element_count', 0)
    return total /len(result_list)


def calculate_unexpected(result_json):
    # Initialize a variable to store the total
    total = 0
    # Convert the 'result_json' column value to a list of dictionaries
    result_list = json.loads(result_json).get('result_json', [])
    # Iterate through the list and calculate the total
    for item in result_list:
        result = item.get('result', {})
        total += result.get('unexpected_count', 0)
    return total /len(result_list)

def navWithLogo():
    col1, col2, col3 = st.columns((4,2,4))
    with col2:
        st.image('static/SnowDQ-LOGO.png', use_column_width=True)
    st.markdown('<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">', unsafe_allow_html=True)

    html_code = """
    <div style="display: flex; justify-content: center; align-items: center; height: 100px;">
    <nav class="navbar navbar-expand-lg navbar-light" style="background-color: #e6ecf3; border-radius: 10px; width: 100%;">
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarNavAltMarkup" aria-controls="navbarNavAltMarkup" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="navbarNavAltMarkup">
            <div class="navbar-nav mx-auto">
                <a class="nav-item nav-link" href="/" target="_self">Home</a>
                <a class="nav-item nav-link" href="/Expectations" target="_self">Expectation</a>
                <a class="nav-item nav-link" href="/Suites" target="_self">Suites</a>
            </div>
        </div>
    </nav>
</div>

    """
    col1, col2, col3 = st.columns((3,4,3))
    with col2:
        st.markdown(html_code, unsafe_allow_html=True)


def delete_callback(selected_row):
    with stylable_container(key="container_with_border",css_styles="""{z-index:9999999;
                                                                        background-color: #e6ecf3;
                                                                        position: fixed;
                                                                        top: 40%;
                                                                        right: 45%;
                                                                        width: 25%;
                                                                        height: 30%;
                                                                        border: 3px solid blue;
                                                                        border: 1px solid rgba(49, 51, 63, 0.2);
                                                                        border-radius: 0.5rem;
                                                                        padding: calc(1em - 1px)}"""):
        with st.container():
            st.subheader("**Confirm Delete**")
            st.write("**Are you want to delete a record?**")
            submit_button = st.button("Yes")
            cancel_button = st.button("No")
            if submit_button:
                print("inside submit button with:", selected_row)
                delete_project(selected_row)
                st.session_state["del_button"] = False
                st.session_state["optionButton"] = False
                st.cache_data.clear()  #TODO Need to find alternate solution
            if cancel_button:
                st.session_state["del_button"] = False
                st.session_state["optionButton"] = False
                


def optionContainer(optionButton,selected_row, projectId):
    with stylable_container(key="container_with_border",
                            css_styles="""{z-index: 9999999;
                                        background-color: #e6ecf3;
                                        position: fixed;
                                        top: 20%;
                                        right: 65%;
                                        width: 15%;
                                        height: 25%;
                                        border: 3px solid blue;
                                        border: 1px solid rgba(49, 51, 63, 0.2);
                                        border-radius: 0.5rem;
                                        padding: calc(1em - 1px);}"""):
                    
        edit = st.button("Edit Project", key=f"edit_button_{optionButton}")
        delete = st.button("Delete Project", key=f"delete_button_{optionButton}")
        suite = st.button("Add/Remove Suite", key=f"suite_button_{optionButton}")

        if "edit_button" not in st.session_state:
            st.session_state["edit_button"] = False

        if "del_button" not in st.session_state:
            st.session_state["del_button"] = False

        if "suite_button" not in st.session_state:
            st.session_state["suite_button"] = False

        if edit:
            st.session_state["edit_button"] = not st.session_state["edit_button"]

        if delete:
            st.session_state["del_button"] = not st.session_state["del_button"]

        if suite:
            st.session_state["suite_button"] = not st.session_state["suite_button"]

    if st.session_state["edit_button"] == True:
        # pass
        edit_callback(selected_row)
    if st.session_state["del_button"] == True:
        # pass
        delete_callback(selected_row)
    if st.session_state["suite_button"] == True:
        # pass
        addSuite_callback(selected_row, projectId)


def edit_callback(selected_row):
    with stylable_container(key="container_with_border",
                            css_styles="""{background-color: #e6ecf3;
                            position: fixed;
                            top:50px;
                            right: 0;
                            width: 450px;
                            z-index:9999;
                            height: 100%;
                            border: 3px solid blue;
                            border: 1px solid rgba(49, 51, 63, 0.2);
                            border-radius: 0.5rem;
                            padding: calc(1em - 1px)}"""):
        form = st.form(key="my_form")
        form.write("")
        form.write("")
        form.write("**Update Project**")
        name = form.text_input(label="Enter Project Name", value=selected_row, disabled=True )
        # disabled: bool = False,
        form.write("Description")
        description = form.text_area(label="Enter Description")
        form.write("")
        form.write("")
        form.write("")
        submit_button = form.form_submit_button("Update")
        cancel_button = form.form_submit_button(label="Cancel")
        if submit_button:
            update_project(selected_row, description)
            # query = f"""UPDATE dq_project SET name = '{name}', description = '{desc}' WHERE ID={l[d]}"""
            st.session_state["optionButton"] = False
            st.session_state["edit_button"] = False
            st.cache_data.clear()  #TODO Need to find alternate solution
            st.experimental_rerun()
        if cancel_button:
            st.session_state["optionButton"] = False
            st.session_state["edit_button"] = False
            st.experimental_rerun()



def addSuite_callback(selected_row, projectId):
    with stylable_container(key="container_with_border",
                            css_styles="""{
                            background-color: #e6ecf3;
                            position: fixed;
                            top:25px;
                            right: 0;
                            width: 750px;
                            z-index:9999;
                            height: 100%;
                            border: 3px solid blue;
                            border: 1px solid rgba(49, 51, 63, 0.2);
                            border-radius: 0.5rem;
                            padding: calc(1em - 1px)}"""):
        
        with st.container():
            group_id = []
            added_suites = []
            suitsDf = load_data(st.secrets.DQ_TABLE.SUITE)
            suitsDf['MODIFIED_DATE'] = pd.to_datetime(suitsDf['MODIFIED_DATE'])
            suitsDf = suitsDf.sort_values(by='MODIFIED_DATE', ascending=False)

            if group_id not in st.session_state:
                st.session_state.group_id = []

            if added_suites not in st.session_state:
                st.session_state.added_suites = []
                

            search = st.text_input("Search Expectation",placeholder="Search suite by name")
            if search:
                suitsDf = suitsDf[suitsDf['NAME'].str.contains(search, case=False)]                        
            
            st.subheader("Add/Remove Suites")
            for index, row in suitsDf.iterrows():
                check = st.checkbox(suitsDf["NAME"][index], key=f"check_{index}")
                if check:
                    st.session_state.group_id.append(suitsDf["ID"][index])
                    st.session_state.added_suites.append(suitsDf["NAME"][index])


            st.write(st.session_state.added_suites)
            print("group_id>>>>>>>>>", st.session_state.group_id)
            save = st.button("Save", key=f"save_{1}")
            cancel = st.button("Cancel", key=f"cancel_{2}")

            if save:
                for gid in st.session_state.group_id:
                    add_suite(projectId, gid)
                st.session_state["suite_button"] = False
                st.session_state["optionButton"] = False
                st.experimental_rerun()
            if cancel:
                st.session_state["suite_button"] = False
                st.session_state["optionButton"] = False
                st.experimental_rerun()



def createProjectcontainer():
    with stylable_container(key="container_with_border",
                            css_styles="""{background-color: #e6ecf3;
                            position: fixed;
                            top:50px;
                            right: 0;
                            width: 450px;
                            z-index:9999;
                            height: 100%;
                            border: 3px solid blue;
                            border: 1px solid rgba(49, 51, 63, 0.2);
                            border-radius: 0.5rem;
                            padding: calc(1em - 1px)}"""):
        with st.container():
            st.write("**Add New Project**")
            name = st.text_input(label='Enter Project Name',placeholder='Enter Project Name')
            st.write("Description")
            description = st.text_area(label='Enter Description',placeholder='Enter Description')
            st.write("")
            st.write("")
            st.write("")

            submit_button = st.button('Create Project')
            cancel_button = st.button('Cancel')
            if submit_button:
                print("submit")
                create_project(name, description)
                st.cache_data.clear()  #TODO Need to find alternate solution
                st.session_state['createProject']  = False
                st.experimental_rerun()
            if cancel_button:
                st.session_state['createProject']  = False
                st.experimental_rerun()